scalding-serialization-macros/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/SealedTraitLike.scala

```diff
@@ -0,0 +1,323 @@
+            """)
+        }
+    }.get
+  }
+
+  def put(c: Context)(inputStream: c.TermName, element: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val innerArg = freshT("innerArg")
+    subData.foldLeft(Option.empty[Tree]) {
+      case (optiExisting, (idx, tpe, optiTBuf)) =>
+        val commonPut: Tree = optiTBuf.map { tBuf =>
+          q"""val $innerArg: $tpe = $element.asInstanceOf[$tpe]
+              ${tBuf.put(inputStream, innerArg)}
+              """
+        }.getOrElse[Tree](q"()")
+
+        optiExisting match {
+          case Some(s) =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            } else {
+              $s
+            }
+            """)
+          case None =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            }
+            """)
+        }
+    }.get
+  }
+
+  def length(c: Context)(element: c.Tree)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): CompileTimeLengthTypes[c.type] = {
+    import CompileTimeLengthTypes._
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val prevSizeData = subData.foldLeft(Option.empty[Tree]) {
+      case (optiTree, (idx, tpe, tBufOpt)) =>
+
+        val baseLenT: Tree = tBufOpt.map{ tBuf =>
+          tBuf.length(q"$element.asInstanceOf[$tpe]") match {
+            case m: MaybeLengthCalculation[_] =>
+              m.asInstanceOf[MaybeLengthCalculation[c.type]].t
+
+            case f: FastLengthCalculation[_] =>
+              q"""_root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(${f.asInstanceOf[FastLengthCalculation[c.type]].t})"""
+
+            case _: NoLengthCalculationAvailable[_] =>
+              return NoLengthCalculationAvailable(c)
+            case const: ConstantLengthCalculation[_] =>
+              q"""_root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.ConstLen(${const.toInt})"""
+            case e => sys.error("unexpected input to union length code of " + e)
+          }
+        }.getOrElse(q"_root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(1)")
+        val tmpPreLen = freshT("tmpPreLen")
+
+        val lenT = q"""
+        val $tmpPreLen: _root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.MaybeLength  = $baseLenT
+
+        ($tmpPreLen match {
+          case _root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.ConstLen(l) =>
+            _root_.com.twitter.scalding.serialization.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(l + 1)
```

Comment:
why to we always jump out to dynamic here? If you did:
```scala
sealed trait Or
case class Left(left: Int) extends Or
case class Right(right: Int) extends Or
```
You'd have a constant size of `5` bytes. Just not worth it to do that optimization yet for what might be corner case?